gloo/transport/uv/context.cc

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright (c) 2019-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gloo/transport/uv/context.h>
+
+#include <gloo/common/error.h>
+#include <gloo/common/logging.h>
+#include <gloo/transport/uv/device.h>
+#include <gloo/transport/uv/pair.h>
+#include <gloo/transport/uv/unbound_buffer.h>
+
+namespace gloo {
+namespace transport {
+namespace uv {
+
+using count_t = PendingOpCount::count_t;
+
+ContextMutator::ContextMutator(Context& context, size_t slot, size_t rank)
+    : lock_(context.m_),
+      context_(context),
+      slot_(slot),
+      rank_(rank),
+      it_(context_.remotePendingOp_.find(slot)) {}
+
+ContextMutator::~ContextMutator() {
+  if (it_ != context_.remotePendingOp_.end() && it_->second.empty()) {
+    context_.remotePendingOp_.erase(it_);
+  }
+}
+
+count_t ContextMutator::getRemotePendingRecv() {
+  if (it_ == context_.remotePendingOp_.end()) {
+    return 0;
+  }
+  return it_->second.getRecv(rank_);
+}
+
+count_t ContextMutator::getRemotePendingSend() {
+  if (it_ == context_.remotePendingOp_.end()) {
+    return 0;
+  }
+  return it_->second.getSend(rank_);
+}
+
+count_t ContextMutator::updateRemotePendingRecv(count_t v) {
+  auto it = insertIfNotExists();
+  return it->second.updateRecv(rank_, v);
```

Comment:
where did we use `lock_`?

Reply:
When the context mutator is constructed it acquires the context lock and when it is destructed it releases it. This means it can only exist if the lock is held.